### ISSUE TITLE
Implement SortedSet

### ISSUE THREAD

**Issue by honzajavorek** [honzajavorek/redis-collections#4]
- http://redis.io/commands#sorted_set
- http://code.activestate.com/recipes/576694/

**Comment by bbayles** [comment]
I've been thinking about how to expose the Redis sorted set data type to Python. Currently I feel emulating `OrderedSet` might not be the best way, but I have another idea.

---

The Python `collections.Counter` type is great, but one thing I don't like about it is how it (a) lets you store arbitrary values, (b) punishes you for doing that:

```python
>>> import collections
>>> C = collections.Counter('abbccc')
>>> C['d'] = 'four'
>>> C.most_common()
Traceback (most recent call last):
  File "<pyshell#8>", line 1, in <module>
    C.most_common()
  File "/home/bo/Code/virtualenvs/redis_collections3/lib/python3.5/collections/__init__.py", line 547, in most_common
    return sorted(self.items(), key=_itemgetter(1), reverse=True)
TypeError: unorderable types: str() < int()
>>> list(C.elements())
Traceback (most recent call last):
  File "<pyshell#10>", line 1, in <module>
    list(C.elements())
TypeError: 'str' object cannot be interpreted as an integer
```

What might be a useful collection is what I'm thinking of as `ZCounter`. It would be backed by the [sorted set Redis type](http://redis.io/commands#sorted_set), and would not aim to emulate `collections.Counter` faithfully. Instead, it would restrict values to things that can be cast to `float`. 

I'm imagining using this collection like this:
```python
>>> from redis_collections import ZCounter
>>> zc = ZCounter({'elem_2': 200, 'elem_3': 300, 'elem_4': 400})
>>> zc['elem_1'] = 100  # ZADD for __setitem__
>>> zc['elem_1']  # ZSCORE for __getitem__
100.0
>> len(zc)  # ZCARD for __len__
4
>>> zc.get_members(start=200, stop=400, reverse=True)  # ZREVRANGEBYSCORE
[('elem_3', 300), ('elem_2', 200)]
>>> zc[1:-1]  # ZRANGE for slicing by rank
[('elem_3', 200), ('elem_2', 300)]
>>> zc.index('elem_1')
0
>>> zc['elem_1'] = 'value'
...
ValueError: could not convert string to float: 'value'
```

I kind of want to allow arbitrary keys, which would make the [ZRANGEBYLEX](http://redis.io/commands/zrangebylex) and similar commands unusable.

Still thinking about this; suggestions welcome.

**Comment by honzajavorek** [comment]
What is the reasoning behind using the Z prefix instead of a sensible name suggesting how the structure works? Is it a convention from somewhere else?

**Comment by bbayles** [comment]
I had ZCounter in my head because of how Redis uses `Z` in front of the SortedSet command names. Better name suggestions of course welcome.

**Comment by honzajavorek** [comment]
Ah, good then. I didn't realise the Redis convention.

**Comment by bbayles** [comment]
First collection with the Redis Sorted set is merged in #81. I'm interested in also experimenting with the Redis [Geo type](http://redis.io/commands/geoadd), which is implemented with sorted set.
